pkg/controller/podautoscaler/replica_calculator.go

```diff
@@ -241,20 +241,30 @@ func (c *ReplicaCalculator) GetObjectMetricReplicas(currentReplicas int32, targe
 	}
 
 	usageRatio := float64(utilization) / float64(targetUtilization)
-	if math.Abs(1.0-usageRatio) <= c.tolerance {
-		// return the current replicas if the change would be too small
-		return currentReplicas, utilization, timestamp, nil
-	}
-
-	readyPodCount, err := c.getReadyPodsCount(namespace, selector)
+	replicaCount, timestamp, err = c.getUsageRatioReplicaCount(currentReplicas, usageRatio, namespace, selector)
+	return replicaCount, utilization, timestamp, err
+}
 
-	if err != nil {
-		return 0, 0, time.Time{}, fmt.Errorf("unable to calculate ready pods: %s", err)
+// getUsageRatioReplicaCount calculates the desired replica count based on usageRatio and ready pods count.
+// For currentReplicas=0 doesn't take into account ready pods count and tolerance to support scaling to zero pods.
+func (c *ReplicaCalculator) getUsageRatioReplicaCount(currentReplicas int32, usageRatio float64, namespace string, selector labels.Selector) (replicaCount int32, timestamp time.Time, err error) {
+	if currentReplicas != 0 {
+		if math.Abs(1.0-usageRatio) <= c.tolerance {
+			// return the current replicas if the change would be too small
+			return currentReplicas, timestamp, nil
+		}
+		readyPodCount := int64(0)
+		readyPodCount, err = c.getReadyPodsCount(namespace, selector)
+		if err != nil {
+			return 0, time.Time{}, fmt.Errorf("unable to calculate ready pods: %s", err)
+		}
+		replicaCount = int32(math.Ceil(usageRatio * float64(readyPodCount)))
+	} else {
+		// Scale to zero or n pods depending on usageRatio
+		replicaCount = int32(math.Ceil(usageRatio))
```

Comment:
We have a use case in which we need to scale worker deployments from min=0 to min=100 (for example) based on the SQS queue length. The worker deployments works on the SQS queue.

Do you see this logic working with AWS SQS? SQS does not have a single metric to give the usageRatio because of the long polling of the queue by the consumers, the `ApproximateMessage` may be showing 0 all the time.

Reply:
@alok87 , personally I use this code with RabbitMQ and Prometheus. I have an heuristic of worker utilization which depends on the number of messages in the queue and the number of workers.

Reply:
@DXist, sorry but wasn't clear to me which Prometheus exporter for RabbitMQ you did use on tests, was [prometheus_rabbitmq_exporter](https://github.com/deadtrickster/prometheus_rabbitmq_exporter) or [rabbitmq_exporter](https://github.com/kbudde/rabbitmq_exporter). Moreover, what was the metric used on that? In a rollback scenario which you doesn't have scaling HPA to/from zero feature available, it'll still work? Does it scale to one instead of zero?

Reply:
@fellippe-mendonca, I use the second exporter which is configured in [rabbitmq-ha helm chart](https://github.com/helm/charts/blob/master/stable/rabbitmq-ha/values.yaml#L465)

I have 2 metrics - worker utilisation and ingress/egress message rate ratio. [Here](http://ix.io/1Okl) is a Helm template of PrometheusRule object I've configured.

If you want to rollback the cluster to version without scale-to-zero support you have to update HPA objects explicitly before the rollback. See release notes to this PR.

Reply:
Thanks for the information @DXist, I'll try this approach with an application I'm running with RabbitMQ that doesn't scale well with CPU. I was wondering to use average response time, but I think that ingress/egress rate ratio might be better! Hope this new functionality come to k8s ASAP and further for GKE.